src/223/main/kotlin/org/rust/ide/inspections/import/RsReferenceImporter.kt

```diff
@@ -0,0 +1,45 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.inspections.import
+
+import com.intellij.codeInsight.daemon.ReferenceImporter
+import com.intellij.openapi.editor.Editor
+import com.intellij.psi.PsiFile
+import org.rust.ide.inspections.RsUnresolvedReferenceInspection
+import org.rust.ide.inspections.shouldIgnoreUnresolvedReference
+import org.rust.ide.settings.RsCodeInsightSettings
+import org.rust.lang.core.psi.RsFile
+import org.rust.lang.core.psi.RsMethodCall
+import org.rust.lang.core.psi.RsPath
+import org.rust.lang.core.psi.ext.RsElement
+import java.util.function.BooleanSupplier
+
+class RsReferenceImporter : ReferenceImporter {
+    override fun autoImportReferenceAtCursor(editor: Editor, file: PsiFile): Boolean = false
+
+    @Suppress("UnstableApiUsage")
+    override fun computeAutoImportAtOffset(editor: Editor, file: PsiFile, offset: Int, allowCaretNearReference: Boolean): BooleanSupplier? {
+        val reference = file.findReferenceAt(offset) ?: return null
+        val element = reference.element as? RsElement ?: return null
+
+        val context = when (element) {
+            is RsPath -> RsUnresolvedReferenceInspection.processPath(element)?.context
+            is RsMethodCall -> AutoImportFix.findApplicableContext(element)
+            else -> return null
+        }
+        if (context == null || context.candidates.size != 1) return null
+
+        if (element.shouldIgnoreUnresolvedReference()) return null
+
+        return BooleanSupplier {
+            AutoImportFix(element, context).invoke(element.project)
```

Comment:
Direction action invocation leads to a separate entry in undo log, so Ctrl+Z undoes auto-import instead of actual typing that is extremely annoying:
https://user-images.githubusercontent.com/3221931/201402110-7511fd56-00f6-4857-a4ef-496b5e1fab6d.mp4

```suggestion
            runUndoTransparentWriteAction {
                context.candidates.first().import(element)
            }
```

Reply:
I agree, but looks like `runUndoTransparentWriteAction` doesn't apply formatting (new import can be inserted without new line), so I propose fixing it in a separate PR (anyway this behaviour is present in 2022.2 as well)

Reply:
Okay

Reply:
#9737